File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/between/BetweenEnvelopeTest.kt
package org.wyrdsekai.app.engine.between

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class BetweenEnvelopeTest {

    private val crypto = TestCryptoProvider()

    @Test
    fun signingDataFormatMatchesSpec() {
        val payload = buildJsonObject { put("type", "headline") }
        val data = BetweenEnvelope.signingData("node-1", "node-2", 1234567890L, payload)
        val str = data.decodeToString()
        // Format: src:dst:ts:payload
        assertTrue(str.startsWith("node-1:node-2:1234567890:"))
        assertTrue(str.contains("\"type\":\"headline\""))
    }

    @Test
    fun signingDataUsesAsterikForBroadcast() {
        val payload = JsonPrimitive("test")
        val data = BetweenEnvelope.signingData("node-1", null, 100L, payload)
        assertTrue(data.decodeToString().contains("node-1:*:100:"))
    }

    @Test
    fun createAndSerialize() {
        val identity = NodeIdentity.generate(crypto)
        val payload = buildJsonObject { put("type", "headline"); put("summary", "all is well") }

        val envelope = BetweenEnvelope.create("node-1", null, payload, identity)

        assertEquals(1, envelope.v)
        assertEquals("node-1", envelope.src)
        assertEquals(null, envelope.dst)
        assertTrue(envelope.sig.isNotEmpty())
        assertTrue(envelope.ts > 0)

        // Round-trip serialization
        val bytes = envelope.toBytes()
        val restored = BetweenEnvelope.fromBytes(bytes)
        assertEquals(envelope.src, restored.src)
        assertEquals(envelope.dst, restored.dst)
        assertEquals(envelope.ts, restored.ts)
        assertEquals(envelope.sig, restored.sig)
    }

    @Test
    fun verifyWithTestCrypto() {
        val identity = NodeIdentity.generate(crypto)
        val payload = buildJsonObject { put("test", true) }

        val envelope = BetweenEnvelope.create("node-1", "node-2", payload, identity)
        assertTrue(envelope.verify(identity.publicKey, crypto))
    }

    @Test
    fun envelopeVersionIsOne() {
        val identity = NodeIdentity.generate(crypto)
        val envelope = BetweenEnvelope.create("n1", null, JsonPrimitive("p"), identity)
        assertEquals(1, envelope.v)
    }
}

File: clients/rn/__tests__/engine/between-envelope.test.ts
import {
  signingData, createEnvelope, verifyEnvelope,
  envelopeToBytes, envelopeFromBytes,
} from '../../src/engine/between/BetweenEnvelope';
import { generateIdentity, TestCryptoProvider } from '../../src/engine/between/NodeIdentity';

describe('BetweenEnvelope', () => {
  const crypto = new TestCryptoProvider();

  it('signing data format matches spec', () => {
    const data = signingData('node-1', 'node-2', 1234567890, { type: 'headline' });
    const str = new TextDecoder().decode(data);
    expect(str).toContain('node-1:node-2:1234567890:');
    expect(str).toContain('"type":"headline"');
  });

  it('signing data uses asterisk for broadcast', () => {
    const data = signingData('node-1', null, 100, 'test');
    const str = new TextDecoder().decode(data);
    expect(str).toContain('node-1:*:100:');
  });

  it('create and serialize', () => {
    const identity = generateIdentity(crypto);
    const payload = { type: 'headline', summary: 'all is well' };

    const envelope = createEnvelope('node-1', null, payload, identity, crypto);

    expect(envelope.v).toBe(1);
    expect(envelope.src).toBe('node-1');
    expect(envelope.dst).toBeNull();
    expect(envelope.sig.length).toBeGreaterThan(0);
    expect(envelope.ts).toBeGreaterThan(0);

    // Round-trip serialization
    const bytes = envelopeToBytes(envelope);
    const restored = envelopeFromBytes(bytes);
    expect(restored.src).toBe(envelope.src);
    expect(restored.dst).toBe(envelope.dst);
    expect(restored.ts).toBe(envelope.ts);
    expect(restored.sig).toBe(envelope.sig);
  });

  it('verify with test crypto', () => {
    const identity = generateIdentity(crypto);
    const envelope = createEnvelope('node-1', 'node-2', { test: true }, identity, crypto);
    expect(verifyEnvelope(envelope, identity.publicKey, crypto)).toBe(true);
  });

  it('envelope version is 1', () => {
    const identity = generateIdentity(crypto);
    const envelope = createEnvelope('n1', null, 'p', identity, crypto);
    expect(envelope.v).toBe(1);
  });
});
